### PR TITLE
Use timestamp instead of datetime to achieve faster cookie expiration in CookieJar

### PR DESCRIPTION
**Pull request by Rongronggg9**
## What do these changes do?

Use timestamp instead of `datetime` to achieve faster cookie expiration in `CookieJar`

## Are there changes in behavior for the user?

No.

## Related issue number

#7583
https://github.com/aio-libs/aiohttp/pull/7819#issuecomment-1806933086

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by Rongronggg9** [comment]
The test failure is Windows-specific. See also https://stackoverflow.com/questions/46133223/maximum-value-of-timestamp

File: aiohttp/cookiejar.py [hunk 1]
```diff
@@ -125,13 +134,8 @@ def clear(self, predicate: Optional[ClearCookiePredicate] = None) -> None:
...
 
-        next_expiration = min(self._expirations.values(), default=self._max_time)
-        try:
-            self._next_expiration = next_expiration.replace(
-                microsecond=0
-            ) + datetime.timedelta(seconds=1)
-        except OverflowError:
-            self._next_expiration = self._max_time
+        next_expiration = min(self._expirations.values(), default=self.MAX_TIME)
+        self._next_expiration = min(next_expiration + 1, self.MAX_TIME)
```
Comment (Dreamsorcerer) [inline aiohttp/cookiejar.py]: This seems to be about twice as fast:
```suggestion
        self._next_expiration = min(*self._expirations.values(), self.MAX_TIME-1) + 1
```

File: aiohttp/helpers.py [hunk 2]
```diff
@@ -530,11 +530,9 @@ def is_ip_address(host: Optional[Union[str, bytes, bytearray, memoryview]]) -> b
     return is_ipv4_address(host) or is_ipv6_address(host)
 
 
-def next_whole_second() -> datetime.datetime:
+def next_whole_second() -> int:
     """Return current time rounded up to the next whole second."""
-    return datetime.datetime.now(datetime.timezone.utc).replace(
-        microsecond=0
-    ) + datetime.timedelta(seconds=0)
+    return ceil(time.time())
```
Comment (Dreamsorcerer) [inline aiohttp/helpers.py]: I'm also thinking that we can just remove this function as it's been greatly simplified and is only used twice.

**Comment by bdraco** [comment]
```python
import timeit
from http import cookies

from yarl import URL

from aiohttp import CookieJar


def filter_large_cookie_jar():
    """Filter out large cookies from the cookie jar."""
    jar = CookieJar()
    c = cookies.SimpleCookie()
    domain_url = URL("http://maxagetest.com/")
    other_url = URL("http://otherurl.com/")

    for i in range(5000):
        cookie_name = f"max-age-cookie{i}"
        c[cookie_name] = "any"
        c[cookie_name]["max-age"] = 60
        c[cookie_name]["domain"] = "maxagetest.com"
    jar.update_cookies(c, domain_url)
    assert len(jar) == 5000
    assert len(jar.filter_cookies(domain_url)) == 5000
    assert len(jar.filter_cookies(other_url)) == 0

    filter_domain = timeit.timeit(lambda: jar.filter_cookies(domain_url), number=1000)
    filter_other_domain = timeit.timeit(
        lambda: jar.filter_cookies(other_url), number=1000
    )
    print(f"filter_domain: {filter_domain}")
    print(f"filter_other_domain: {filter_other_domain}")


filter_large_cookie_jar()
```

Basic benchmark to test against (can probably be improved a bit), although I don't think this PR will affect it very much

File: aiohttp/cookiejar.py (hunk 1 above)
Comment (Rongronggg9) [reply aiohttp/cookiejar.py]: It leads to an error if `self._expirations` is empty: `TypeError: 'int' object is not iterable`.
I read the code twice and found `self._next_expiration` never actually used anywhere, may we just remove it?

Comment (Rongronggg9) [reply aiohttp/cookiejar.py]: The usage of `self._next_expiration` has been removed since e65f1a9eba0ebde848bf0e50aa9be4753334b3c8. If it is intended to be a shortcut of `_do_expiration`, it should be added back.

Comment (bdraco) [reply aiohttp/cookiejar.py]: @Rongronggg9 please see the discussion in https://github.com/aio-libs/aiohttp/issues/7790

Comment (Rongronggg9) [reply aiohttp/cookiejar.py]: @bdraco Thanks for your information! I didn't notice that.
So for the time being I decided to just keep it :)

Comment (Dreamsorcerer) [reply aiohttp/cookiejar.py]: You can just unpack into a tuple, and it'll work with 0-length lists:
`self._next_expiration = min((*self._expirations.values(), self.SUB_MAX_TIME)) + 1`

Comment (Dreamsorcerer) [reply aiohttp/cookiejar.py]: The if statement seems to be a tiny bit faster though, so may as well stick with that.

File: aiohttp/cookiejar.py [hunk 3]
```diff
@@ -52,9 +55,24 @@ class CookieJar(AbstractCookieJar):
...
+        int(datetime.datetime.max.replace(tzinfo=datetime.timezone.utc).timestamp()) - 1
+    )
+    try:
+        calendar.timegm(time.gmtime(MAX_TIME))
+    except OSError:
+        # Hit the maximum representable time on Windows
+        # https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/localtime-localtime32-localtime64
+        MAX_TIME = calendar.timegm((3000, 12, 31, 23, 59, 59, -1, -1, -1))
+        # The below line should never raise an exception
+        time.gmtime(MAX_TIME)
```
Comment (Dreamsorcerer) [inline aiohttp/cookiejar.py]: Why is this line here? Should we just remove it?

File: aiohttp/cookiejar.py [hunk 4]
```diff
@@ -81,14 +99,8 @@ def __init__(
...
         self._treat_as_secure_origin = treat_as_secure_origin
-        self._next_expiration = next_whole_second()
-        self._expirations: Dict[Tuple[str, str, str], datetime.datetime] = {}
-        # #4515: datetime.max may not be representable on 32-bit platforms
-        self._max_time = self.MAX_TIME
-        try:
-            self._max_time.timestamp()
-        except OverflowError:
-            self._max_time = self.MAX_32BIT_TIME
+        self._next_expiration: Union[int, float] = ceil(time.time())
```
Comment (Dreamsorcerer) [inline aiohttp/cookiejar.py]:
```suggestion
        self._next_expiration: float = ceil(time.time())
```
int is compatible with float

File: aiohttp/cookiejar.py [hunk 5]
```diff
@@ -81,14 +99,8 @@ def __init__(
...
+        self._expirations: Dict[Tuple[str, str, str], Union[int, float]] = {}
```
Comment (Dreamsorcerer) [inline aiohttp/cookiejar.py]:
```suggestion
        self._expirations: Dict[Tuple[str, str, str], float] = {}
```

File: aiohttp/cookiejar.py [hunk 6]
```diff
@@ -148,7 +158,7 @@ def _do_expiration(self) -> None:
         self.clear(lambda x: False)
 
     def _expire_cookie(
-        self, when: datetime.datetime, domain: str, path: str, name: str
+        self, when: Union[int, float], domain: str, path: str, name: str
```
Comment (Dreamsorcerer) [inline aiohttp/cookiejar.py]:
```suggestion
        self, when: float, domain: str, path: str, name: str
```

**Review by Dreamsorcerer — APPROVED** [review]

File: aiohttp/cookiejar.py (hunk 3 above)
Comment (Rongronggg9) [reply aiohttp/cookiejar.py]: We assume that OSError will only occur on Windows, without really verifying the current OS. The line is to ensure if there is any other (OS, arch) that raises OSError, the Windows-specific max time won't be too large (we should receive bug reports if it occurs).

Comment (Dreamsorcerer) [reply aiohttp/cookiejar.py]: But, if it does happen, then the whole library fails on import, even if the user would never have received a value anywhere near the MAX_TIME and wouldn't have experienced any issues at runtime.

Comment (Rongronggg9) [reply aiohttp/cookiejar.py]: OK then. I've removed the line

**Comment by Dreamsorcerer** [comment]
#7837
